[CIR] Add support for binary arithmetic operations on complex numbers

Lancern opened the pull request:
This PR adds support for complex number binary arithmetic operations. Specifically, CIRGen for the following operations on complex numbers are added:

  - Binary arithmetic operations: add, sub, mul, and div.
  - ~Unary arithmetic operations: plus, minus, and conjugate.~

A new operation `cir.complex.binop` is added to represent binary operations on complex number operands. The new operation contains complex-number-specific attributes for correct lowering to LLVM IR.

~I'll add LLVM IR lowering in later PRs as this PR is already large enough.~ This PR also includes LLVM IR lowering support. It introduces two new operations `cir.complex.real` and `cir.complex.imag` to aid for LLVM IR lowering. These two new operations extract the real and imaginary part of a complex value, which is useful during lowering prepare, during which phase the complex arithmetic operations are broken down to a series of scalar arithmetic operations on the real and imaginary parts of the complex operands.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -1258,6 +1258,61 @@ def ComplexImagPtrOp : CIR_Op<"complex.imag_ptr", [Pure]> {
+}
+
+def ComplexBinOp : CIR_Op<"complex.binop",
+    [Pure, SameTypeOperands, SameOperandsAndResultType]> {
```
I'm a bit on the fence here, my understanding is that the range stuff can only be used with mul and div, right? If so, perhaps add and sub should just be regular binops? We'd need to (1) change the existing binop verifier to choke on complex types out of sub and add (2) Make ComplexRangeKind mandatory in this new op, and you can also remove the complex.binop verifier for it.

There are also no tests that check for the range stuff output, can you please add that for each?

bcardosolopes commented on clang/test/CIR/CodeGen/complex-arithmetic.c:
```diff
@@ -0,0 +1,128 @@
+
+double _Complex cd1, cd2;
+int _Complex ci1, ci2;
+
```
Seems like we're missing tests for the unary ops?

Lancern replied:
The tests for unary ops are here: https://github.com/llvm/clangir/pull/730/files/6261dc6010e5cb46cb9f780d7acf11442c8d8554#diff-af00df30674a0105a6f183bf20b1f389d2d586c6552a6a03a04c7f9ea5de41beR97-R128

bcardosolopes reviewed (changes requested):
Thanks for working on this!

Quick reminder: it's preferable to add less CIRGen but add LLVM lowering altogether than not having LLVM support as part of the PR - the rationale is the same as I mentioned before: one might need to change CIRGen again when they go for LLVM lowering because something might have lacked on the design. Fine for this PR, but might be worth changing the approach looking fwd (PRs might land faster because it's also easier for me to review and test the final LLVM output, which I usually do locally in order to build extra trust on the approach).

Lancern commented:
> it's preferable to add less CIRGen but add LLVM lowering altogether than not having LLVM support as part of the PR.

OK I'm changing this PR to draft to add LLVMIR support as well -- the LLVMIR lowering should be OK for review in size. I agree on this rationale as it helps me verify my CIRGen scheme.

Lancern commented:
Updated this PR and added LLVM IR lowering support and tests. Removed support code for unary operations as this PR is already large enough. Unary operations may come later in a future PR.

The update also introduces two new operations `cir.complex.real` and `cir.complex.imag`. These two operations extract the real and imaginary part of a complex value. During lowering prepare, complex binary arithmetic operations are broken down to scalar arithmetic operations on the operands' real and imaginary parts, and these two operations are generated there. For example:

```mlir
%2 = cir.binop add %0, %1 : !cir.complex<!cir.float>
```

will be transformed to the following CIR during lowering prepare:

```mlir
%0.real = cir.complex.real %0
%0.imag = cir.complex.imag %0
%1.real = cir.complex.real %1
%1.imag = cir.complex.imag %1
%res.real = cir.binop add %0.real, %1.real
%res.imag = cir.binop add %0.imag, %1.imag
%2 = cir.complex.create %res.real, %res.imag
```

And the above CIR will be further lowered to LLVM IR during LLVM IR lowering.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -3306,6 +3326,12 @@ LogicalResult BinOp::verify() {
+                    getKind() == mlir::cir::BinOpKind::Sub;
+  if (isa<mlir::cir::ComplexType>(getType()) && !complexOps)
+    return emitError()
+           << "cir.binop can only represent 'add' and 'sub' on complex numbers";
```
one small improvement could be "numbers, use cir.complex.binop instead", or something like that.

bcardosolopes reviewed (approved):
LGTM + nit that can be addressed in a later PR.
